Local development guidelines for 2022

mariusandra opened the pull request:
## Changes

- WIP. Trying to update the existing local development guidelines and merge the Postgres and ClickHouse setups.
- Still many points to go.
- Discussion: I think we need to ruthlessly simplify the number of options we offer users. Currently over the [two](https://posthog.com/handbook/engineering/ee-setup) [docs](https://posthog.com/docs/contribute/developing-locally), we have so many options to choose from that a new contributor will have a hard time deciding what to do. I propose we stick with two: a) everything in docker, b) all services in docker (including postgres), all code locally.

## Checklist
- [ ] Titles are in [sentence case](https://apastyle.apa.org/style-grammar-guidelines/capitalization/sentence-case)
- [ ] Words are spelled using American english
- [ ] I have checked out our [styleguide](https://github.com/PostHog/posthog.com/blob/master/STYLEGUIDE.md)

mariusandra commented:
This is ready for a look.

I've rewritten the guide and kept just one recommended development option:

- Run services (postgres, redis, clickhouse, kafka, zookeeper) in Docker, and run the apps locally

I tried more hours than I'm willing to publicly admit (definitely over a day) to get the "run everything in docker" approach working, but at the end of the day it was way too slow, very flakey, and not useful anyway. The problems with the "run it all in Docker" approach:

- I discovered that unless you have installed all the node and python packages locally, `git commit` will always fail, because it's missing `flake8`, `prettier` and `eslint`. If the user already needs to install node and python and these specific packages, they might as well install the rest of our dependencies.
- It was not a problem to install Django in Docker, but the frontend caused a lot of problems. Mainly regarding `node_modules`. Running `yarn install` when sharing the folder between the host and the container took minutes, and sometimes did not complete. Running `yarn install` during the build wouldn't work, as `volume=..:/code` would override the downloaded modules. Running `yarn install` during the build, caching the modules, and then copying them back to `/code` in the entrypoint script also didn't work. My CPU got stuck at ~400% while it tried to sync the huge folder with localhost... and didn't finish within a few minutes. Considering this is what should be done before the app starts, it's too much.
- What did work was building a container, and not attaching the running code as a volume. However in this approach you must rebuild the container after all changes, which is not practical.

Hence I skipped the approach.

We could bring back "run frontend locally and all the rest in docker", but since you anyway need to install python and its packages, I see little value in only running the frontend locally, and not the rest of the stack.

Twixes commented:
Great timing, just set up everything _but_ PostHog itself on the new hardware, so giving this a go now.

macobo commented on contents/docs/contribute/developing-locally.mdx:
```diff
@@ -9,231 +9,268 @@ showTitle: true
+
+3. Install [Docker Desktop](https://www.docker.com/products/docker-desktop) and give it **at least 4 GB of RAM** (6GB if you can afford it) and 8 CPU cores.
+
+4. Add the following lines to `/etc/hosts`:
```
I have some context:

I don't think this is needed and will cause trouble for people running multiple projects at once.

Instead I have these following env vars set up for myself

```
⟩ export KAFKA_ENABLED=true KAFKA_HOSTS=localhost:9092 KAFKA_URL=kafka://localhost:9092
```

Frustratingly, there's two url ones because plugin-server and app use different ones.

There's a gotcha here around `migrate_clickhouse` though - since clickhouse accesses network over docker, the host needs to be `kafka` there but `localhost` locally. I think we could improve this by better defaults in the app though.

mariusandra replied:
> I don't think this is needed 

Sorry, but things just don't work for me without this... so I do think it's needed. ClickHouse and Kafka are unable to communicate. Probably the same for other Docker on Mac users (please confirm!). I'll try once more without, but happy to add a "skip if non mac" line there.

macobo replied:
I think the reason it didn't work it's because of the `KAFKA_HOSTS` setting being set when `python manage.py migrate_clickhouse` is run.

If we go the route of "sane defaults if DEBUG=1 is set" I think I have a solution for that and can solve this in a follow-up. This would be conditionally setting `KAFKA_HOSTS` to kafka if DEBUG during `migrate_clickhouse` so the kafka engine table is set up correctly :)

macobo commented on contents/docs/contribute/developing-locally.mdx:
````diff
@@ -9,231 +9,268 @@ showTitle: true
 ```
 
-6. Install dev requirements:
+3. Create the virtual environment in current directory called 'env':
````
If we're using pyenv then manually creating venvs is unneeded. You can do this instead:

- `pyenv virtualenv 3.9.9 posthog`
- Use some way to automatically activate that when you cd into the dir - I use https://github.com/eknkc/fish-autoenv

macobo commented on contents/docs/contribute/developing-locally.mdx:
````diff
@@ -9,231 +9,268 @@ showTitle: true
 
-<HiddenSection headingType='h2' title='Using Docker'>
+```bash
+export DEBUG=1
````
If you use https://github.com/eknkc/fish-autoenv like me, you can add these to your .env.fish not to have to type these out every time :)

mariusandra commented:
Replying outside of [this thread](https://github.com/PostHog/posthog.com/pull/2571#discussion_r765829483) for more visibility.

Some of these steps require passing in random envs before running. I want to greatly simplify this. If we 1) agree on "services in docker, apps running locally" being the recommended approach, and 2) deprecate postgres..., then setting `DEBUG=1`, should at least also set these values:

``` bash
export PRIMARY_DB=clickhouse
export DATABASE_URL=postgres://posthog:posthog@localhost:5432/posthog
export KAFKA_ENABLED=true
export KAFKA_HOSTS=kafka:9092
export KAFKA_URL=kafka://kafka:9092
```

Twixes reviewed (commented):
Totally agreed regarding env vars @mariusandra. Setting this up now, having the appropriate env vars has definitely been my biggest pain point, particularly because _some_ scripts in `bin/` assume Homebrew Postgres, while others Docker Postgres – things would be much smoother if this was unified.

I'm not going to green mark this since I made some major changes myself, but I _think_ this looks good and is reasonably nice to follow. As for the changes:
- removed duplicated remnants of the old guide from the bottom
- marked PyCharm and feature flags sections as extras
- clarified PostHog architecture
- clarified steps: first external services, then **preparing** PostHog components, then starting PostHog
- converted `pyenv` to plain `python3` (may be contentious)
- converted `docker-compose` to `docker compose` – only the latter runs natively on ARM, without Rosetta
- changed frontend step from `yarn start` to just `yarn typegen:write` – when going through this, it seemed a bit confusing why `yarn start` and when to stop it, considering that the last step is `bin/start`

Overall things are working pretty smoothly (well, except for the env var complexities mentioned before).

BTW: What about the "Setting up Clickhouse for development" Handbook page? I suppose it should not be needed separately to "Developing locally" in 2022. _And_ there's another seemingly redundant page "4. Developer Workflow".

mariusandra commented:
Thanks @Twixes! Found a few other small tweaks to make, mostly regarding numbering. I'm fine with using `python3` directly from brew, though we might run the risk of this installing e.g. `python 3.10` when we only officially support 3.8 and 3.9. Perhaps there should be a "friendly tip" suggesting `pyenv` if you need to try other versions. Happy to get the instructions simpler and remove yet another dependency though.

I'll add a bunch of other reviewers to this PR now. When reviewing, please keep in mind that the instructions are *descriptive*, describing what you currently need to do, and geared towards *anyone* from the internet who wants to contribute to us. Greatly simplifying some of the steps (e.g. making `PRIMARY_DB=clickhouse` default) is for follow up PRs.

guidoiaquinti commented on contents/docs/contribute/developing-locally.mdx:
```diff
@@ -9,264 +9,282 @@ showTitle: true
+3. Install [Docker Desktop](https://www.docker.com/products/docker-desktop) and in its settings give Docker **at least 4 GB of RAM** (or 6 GB if you can afford it) and at least 4 CPU cores.
 
-2. Make sure you have [PostgreSQL installed](https://www.postgresql.org/download/) and running. We recommend using version 11 or higher. You may also try [Postgres.app](https://postgresapp.com/), but remember to follow the instructions to add `psql` to your `$PATH` if you do.
+4. Append line `127.0.0.1 kafka clickhouse` to `/etc/hosts`. You can do it in one line with:
```
What's the point of doing this? What about using `127.0.0.1` instead of `kafka` and `clickhouse` in the env variables?

mariusandra replied:
I'd be super happy to not need this... but have no time to try all permutations of variables/hosts now. If anyone can report our setup working well on a M1 mac without these steps, either with changed envs or not, please advise.

guidoiaquinti replied:
Let's merge with this and then we can re-evaluate.

fuziontech replied:
this is because of an implementation detail on Kafka side. Basically your local client first connects to the broker asking for members of the kafka cluster it gives you `kafka` as the hostname (because of how docker-compose is configured with that as the name of the service & how the services refer to kafka from within the compose setup). So after your client asks the broker for the connection details it tries to connect to `kafka` host from local. Having `kafka` route to local and into the compose stack is the easiest way to get around that.

fuziontech replied:
Re: ClickHouse I'm not sure why this is required - I haven't had this issue in the past. I'll look into it.

guidoiaquinti commented on contents/docs/contribute/developing-locally.mdx:
````diff
@@ -9,264 +9,282 @@ showTitle: true
-    CREATE USER posthog WITH ENCRYPTED PASSWORD 'posthog';
-    GRANT ALL PRIVILEGES ON DATABASE posthog, posthog_e2e_test TO posthog;
-    ```
+```bash
````
Nit: we should add 4 spaces to all code blocks in the bulleted list

guidoiaquinti commented on contents/docs/contribute/developing-locally.mdx:
```diff
@@ -9,264 +9,282 @@ showTitle: true
-> **Another friendly tip:** You may run into `psycopg2` errors while migrating on a Apple Silicon machine. Try out the steps in this [comment](https://github.com/psycopg/psycopg2/issues/1216#issuecomment-820556849) to resolve this. 
-    
-8. Make sure you have [Yarn installed](https://classic.yarnpkg.com/en/docs/install/):
+If your workstation is ARM-based (e.g. Apple Silicon), the first time your run `pip install` you must pass it custom OpenSSL headers:
```
Ditto

guidoiaquinti commented:
Minor comments but LGTM 👏

pauldambra commented on contents/docs/contribute/developing-locally.mdx:
```diff
@@ -9,264 +9,282 @@ showTitle: true
+
+### 4. Prepare the Django server
+
+1. Install Python 3.8+. You can do so with Homebrew: `brew install python3`. Make sure to always use `python3` instead of `python`, as the latter may point to Python 2.x on some systems.
```
```suggestion
1. Install Python 3.8+. You can do so with Homebrew: `brew install python@3.8` or `brew install python@3.9`.
```

To install a fixed version.

I'd keep `pyenv` because global dependencies can make for confusing bugs but "all the ways to manage python on your machine" can be kept out of scope :)

paolodamico replied:
+1 to `pyenv`, makes things easier (at least for me)

pauldambra reviewed (approved):
much clearer! well written 🙌

paolodamico commented on contents/docs/contribute/developing-locally.mdx:
```diff
@@ -9,264 +9,282 @@ showTitle: true
+To see debug logs (such as ClickHouse queries), add argument `--log-cli-level=DEBUG`.
 
-Run `DEBUG=1 ./bin/start-worker`
+For Cypress end-to-end test, run `bin/e2e-test-runner`. This will temporarily install required dependencies inside the project, spin up a test instance of PostHog, and show you the Cypress interface, from which you'll manually choose tests to run.
```
```suggestion
For Cypress end-to-end test, run `bin/e2e-test-runner`. This will temporarily install required dependencies inside the project, spin up a test instance of PostHog, and show you the Cypress interface, from which you'll manually choose tests to run.

Once you're done, terminate the command with cmd + C. Be sure to wait until the command terminates gracefully so temporary dependencies are removed from `package.json` and you don't commit them accidentally.
```

mariusandra replied:
done!

paolodamico commented on contents/docs/contribute/developing-locally.mdx:
```diff
@@ -9,264 +9,282 @@ showTitle: true
+For Cypress end-to-end test, run `bin/e2e-test-runner`. This will temporarily install required dependencies inside the project, spin up a test instance of PostHog, and show you the Cypress interface, from which you'll manually choose tests to run.
 
-### Running frontend separately (React)
+For frontend tests, all you need is `yarn test`.
```
Do we want to add instructions to run Cypress component tests?

mariusandra replied:
Not really. I think we're [deprecating](https://github.com/PostHog/posthog/issues/6589) them.

paolodamico reviewed (approved):
lgtm! just went through the process myself and A+

paolodamico commented on contents/handbook/engineering/ee-setup.md:
```diff
@@ -4,6 +4,8 @@ sidebar: Handbook
 showTitle: true
 ---
 
+This page is deprecated. Please read [Developing Locally](/docs/contribute/developing-locally) instead.
```
Why don't we delete this instead and just add a redirect?

mariusandra replied:
done!

mariusandra commented:
Plenty of people have had the chance to make their opinions known, so merging this as is, with these extras:
- Settled on saying "python 3.9" instead of "3.8+" to make things cleaner and avoid the "so what do I install then?" question
- Added a link to `pyenv` in case you want a beefier setup than what `brew install python@3.9` gives you.
- Merged in some other changes and suggestions.
- Removed the other, "ee-setup" page and added a redirect.

Let's merge this in now. The next steps would be to radically simplify many of the steps and envs, so that `DEBUG=1` is all you need to set. I'm happy to give someone else a chance work on this :).

fuziontech reviewed (commented):
:shipit:
